ext/signals.c

```diff
@@ -298,22 +312,108 @@ void ddtrace_signals_first_rinit(void) {
             return;
         }
 
-        size_t stack_size = SIGSTKSZ < MIN_STACKSZ ? MIN_STACKSZ : SIGSTKSZ;
-        if ((ddtrace_altstack.ss_sp = malloc(stack_size))) {
-            ddtrace_altstack.ss_size = stack_size;
-            ddtrace_altstack.ss_flags = 0;
-            if (sigaltstack(&ddtrace_altstack, NULL) == 0) {
-                ddtrace_sigaction.sa_flags = SA_ONSTACK;
-                ddtrace_sigaction.sa_handler = ddtrace_sigsegv_handler;
-                sigemptyset(&ddtrace_sigaction.sa_mask);
-                sigaction(SIGSEGV, &ddtrace_sigaction, NULL);
-            }
+        dd_signals_init_async_stack();
+
+        dd_altstack.ss_sp = dd_signal_async_stack;
+        dd_altstack.ss_size = dd_signal_async_stack_size;
+        dd_altstack.ss_flags = 0;
+        if (sigaltstack(&dd_altstack, NULL) == 0) {
+            dd_sigsegv_sigaction.sa_flags = SA_ONSTACK;
+            dd_sigsegv_sigaction.sa_handler = dd_sigsegv_handler;
+            sigemptyset(&dd_sigsegv_sigaction.sa_mask);
+            sigaction(SIGSEGV, &dd_sigsegv_sigaction, NULL);
         }
     }
 }
 
+#if __linux
+static struct sigaction dd_sigint_sigterm_sigaction;
+static struct sigaction dd_sigterm_prev_sigaction;
+static struct sigaction dd_sigint_prev_sigaction;
+
+struct {
+    int sig;
+    siginfo_t si;
+    void *uc;
+} dd_signal_data;
+
+static int dd_call_prev_handler(bool flush) {
+    struct sigaction prev_sigaction = dd_signal_data.sig == SIGINT ? dd_sigint_prev_sigaction : dd_sigterm_prev_sigaction;
+    void *prev_handler = (prev_sigaction.sa_flags & SA_SIGINFO) ? (void *)prev_sigaction.sa_sigaction : (void *)prev_sigaction.sa_handler;
+    if (prev_handler == SIG_IGN) {
+        return 0;
+    }
+
+    if (flush) {
+        ddog_sidecar_flush_traces(&ddtrace_sidecar);
+    }
+
+    if (prev_handler == SIG_DFL) {
+        _exit(0);
+    }
+
+    if (prev_sigaction.sa_flags & SA_SIGINFO) {
+        (*prev_sigaction.sa_sigaction)(dd_signal_data.sig, &dd_signal_data.si, dd_signal_data.uc);
+    } else {
+        (*prev_sigaction.sa_handler)(dd_signal_data.sig);
+    }
+
+    return 0;
+
+}
+
+static int dd_sigterm_cleanup_thread(void *arg) {
+    return dd_call_prev_handler(true);
+}
+
+static void dd_sigint_sigterm_handler(int sig, siginfo_t *si, void *uc) {
+    dd_signal_data.sig = sig;
+    memcpy(&dd_signal_data.si, si, sizeof(*si));
+    dd_signal_data.uc = uc;
+
+    if (ddtrace_sidecar) {
+        // Spawn a thread using clone() to perform sidecar cleanup asynchronously to avoid async unsafeness in the signal handler
```

Comment:
I don't understand. We're supposed to be forcing a flush because the entry point (or a child of it) is shutting down. But this makes a thread... why would this have a realistic chance to complete before the process exits?